galley/pkg/mcp/source.go

```diff
@@ -83,45 +134,117 @@ func (s *source) Stop() {
 }
 
 func (s *source) Apply(c *client.Change) error {
+	_, found := metadata.Types.Lookup(c.TypeURL)
+	if !found {
+		return fmt.Errorf("invalid type: %v", c.TypeURL)
+	}
+	scope.Debugf("received object %+v and length %d", c, len(c.Objects))
+
+	// By default mark all the elements of local cache deleted, if present
+	if _, ok := s.lCache[c.TypeURL]; !ok {
+		s.lCache[c.TypeURL] = make(map[string]*cache)
+	} else {
+		for _, c := range s.lCache[c.TypeURL] {
+			c.ek = resource.Deleted
+		}
+	}
+
 	var errs error
 	for _, o := range c.Objects {
+		if o.TypeURL != c.TypeURL {
+			errs = multierror.Append(errs,
+				fmt.Errorf("type %v mismatch in received object: %v", c.TypeURL, o.TypeURL))
+			continue
+		}
 		e, err := toEntry(o)
 		if err != nil {
 			errs = multierror.Append(errs, err)
+			continue
+		}
+		// If there is an error, ignore the entire batch.
+		// but keep appending errs[] for the batch
+		if errs == nil {
+			lc, ok := s.lCache[o.TypeURL][o.Metadata.Name]
+			if ok {
+				if lc.version == o.Metadata.Version {
+					lc.ek = resource.None
+				} else {
+					lc.prevVer = lc.version
+					lc.version = o.Metadata.Version
+					lc.ek = resource.Updated
+				}
+				lc.entry = &e
+
+			} else {
+				s.lCache[o.TypeURL][o.Metadata.Name] = &cache{
+					version: o.Metadata.Version,
+					ek:      resource.Added,
+					entry:   &e,
+				}
+			}
 		}
+	}
 
-		s.c <- resource.Event{
-			// TODO: ensure FullSync is the correct kind; because we get the full resource I think this is the right thing to do
-			Kind:  resource.FullSync,
-			Entry: e,
+	// If there is an error, the entire batch will be ignored
+	// Hence remove any new adds from the local cache
+	// And restore prev version if there were updates
+	if errs != nil {
+		for key, lc := range s.lCache[c.TypeURL] {
+			if lc.ek == resource.Added {
+				delete(s.lCache[c.TypeURL], key) //this is safe
+			} else if lc.ek == resource.Updated {
+				lc.version = lc.prevVer // We are not worried about the entry as it will be overwritten
+			}
+		}
+	} else {
```

Comment:
Return at the end of this block and get rid of the else, out-denting the entire block.
```go
	if errs != nil {
		for key, lc := range s.lCache[c.TypeURL] {
			// ...
		}
		return errs
	}
```